mlir/lib/Driver/CompilerDriver.cpp

```diff
@@ -567,62 +637,90 @@ LogicalResult QuantumDriverMain(const CompilerOptions &options, CompilerOutput &
     sourceMgr->AddNewSourceBuffer(std::move(moduleBuffer), SMLoc());
     SourceMgrDiagnosticHandler sourceMgrHandler(*sourceMgr, &ctx, options.diagnosticStream);
 
-    OwningOpRef<ModuleOp> op =
+    DefaultTimingManager tm;
+    applyDefaultTimingManagerCLOptions(tm);
+    TimingScope timing = tm.getRootScope();
+
+    TimingScope parserTiming = timing.nest("Parser");
+    OwningOpRef<ModuleOp> mlirModule =
         timer::timer(parseMLIRSource, "parseMLIRSource", /* add_endl */ false, &ctx, *sourceMgr);
-    catalyst::utils::LinesCount::ModuleOp(*op);
-    output.isCheckpointFound = options.checkpointStage == "mlir";
+
+    enum InputType inType = InputType::OTHER;
+    if (mlirModule) {
+        inType = InputType::MLIR;
+        catalyst::utils::LinesCount::ModuleOp(*mlirModule);
+        output.isCheckpointFound = options.checkpointStage == "mlir";
+    }
+    else {
+        llvm::SMDiagnostic err;
+        llvmModule = timer::timer(parseLLVMSource, "parseLLVMSource", false, llvmContext,
+                                  options.source, options.moduleName, err);
+
+        if (!llvmModule) {
+            err.print(options.moduleName.data(), options.diagnosticStream);
+            CO_MSG(options, Verbosity::Urgent, "Failed to parse module as LLVM or MLIR source\n");
+            return failure();
+        }
+        inType = InputType::LLVMIR;
+        output.isCheckpointFound = options.checkpointStage == "llvm_ir";
+        catalyst::utils::LinesCount::Module(*llvmModule);
+    }
+    if (failed(verifyInputType(options, inType))) {
+        return failure();
+    }
+    parserTiming.stop();
 
     // Enzyme always happens after O2Opt. If the checkpoint is O2Opt, enzymeRun must be set to
     // true so that the enzyme pass can be executed.
     bool enzymeRun = options.checkpointStage == "O2Opt";
-    if (op) {
-        enzymeRun = containsGradients(*op);
-        if (failed(runLowering(options, &ctx, *op, output))) {
+
+    bool runAll = options.loweringAction == Action::All;
```

Comment:
```suggestion
    bool runAll = (options.loweringAction == Action::All);
```

Reply:
I always like putting brackets for readability, though there's a chance the formater does not like it. If that's the case ignore this comment : )